Tools/i18n/pygettext.py

```diff
@@ -325,203 +314,81 @@ def add_location(self, filename, lineno, msgid_plural=None, *, is_docstring=Fals
-        spec = self.__options.keywords[self.__curr_keyword]
-        arg_type = spec.get(self.__curr_arg)
-        expect_string_literal = arg_type is not None
-
-        if ttype == tokenize.OP and self.__enclosurecount == 0:
-            if tstring == ')':
-                # We've seen the last of the translatable strings.  Record the
-                # line number of the first line of the strings and update the list
-                # of messages seen.  Reset state for the next batch.  If there
-                # were no strings inside _(), then just ignore this entry.
-                if self.__data:
-                    self.__addentry(self.__data)
-                self.__state = self.__waiting
-                return
-            elif tstring == ',':
-                # Advance to the next argument
-                self.__curr_arg += 1
-                return
 
-        if expect_string_literal:
-            if ttype == tokenize.STRING and is_literal_string(tstring):
-                self.__data[arg_type] += safe_eval(tstring)
-            elif ttype not in (tokenize.COMMENT, tokenize.INDENT, tokenize.DEDENT,
-                               tokenize.NEWLINE, tokenize.NL):
-                # We are inside an argument which is a translatable string and
-                # we encountered a token that is not a string.  This is an error.
-                self.warn_unexpected_token(tstring)
-                self.__enclosurecount = 0
-                self.__state = self.__waiting
-        elif ttype == tokenize.OP:
-            if tstring in '([{':
-                self.__enclosurecount += 1
-            elif tstring in ')]}':
-                self.__enclosurecount -= 1
-
-    def __ignorenext(self, ttype, tstring, lineno):
-        self.__state = self.__waiting
-
-    def __addentry(self, msg, lineno=None, *, is_docstring=False):
-        msgid = msg.get('msgid')
-        if msgid in self.__options.toexclude:
+        if max(spec) >= len(node.args):
+            print(f'*** {self.filename}:{node.lineno}: Expected at least '
+                  f'{max(spec) + 1} positional argument(s) in gettext call, '
+                  f'got {len(node.args)}', file=sys.stderr)
             return
-        if not is_docstring:
-            spec = self.__options.keywords[self.__curr_keyword]
-            if not matches_spec(msg, spec):
+
+        msg_data = {}
+        for position, arg_type in spec.items():
+            arg = node.args[position]
+            if not self._is_string_const(arg):
+                print(f'*** {self.filename}:{arg.lineno}: Expected a string '
+                      f'constant for argument {position + 1}, '
+                      f'got {ast.unparse(arg)}', file=sys.stderr)
                 return
-        if lineno is None:
-            lineno = self.__lineno
-        msgctxt = msg.get('msgctxt')
-        msgid_plural = msg.get('msgid_plural')
+            msg_data[arg_type] = arg.value
+
+        if not matches_spec(msg_data, spec):
```

Comment:
This is no longer needed, because we iterated all `spec` items above.

Reply:
Removed!